extensions/amp-position-observer/0.1/amp-position-observer.js

```diff
@@ -0,0 +1,378 @@
+      return;
+    } else {
+      this.isVisible_ =
+        positionRect.bottom >= (adjustedViewportRect.top + offset);
+      return;
+    }
+  }
+
+  /**
+   * Calculates the current scroll progress as a percentage.
+   * Scroll progress is a decimal between 0-1 and shows progress between
+   * enter and exit, considering ratio and margins.
+   * When a scene becomes visible (enters based on ratio and margins), from
+   * bottom, progress is 0 as it moves toward the top, progress increases until
+   * it becomes exists with 1 from the top.
+   *
+   * Entering from the top gives the reverse values, 1 at enter, 0 at exit.
+   *
+   * @private
+   */
+  updateScrollProgress_(positionRect, adjustedViewportRect) {
+    const totalDurationOffset = (positionRect.height * this.bottomRatio_) +
+        (positionRect.height * this.topRatio_);
+
+    const totalDuration = adjustedViewportRect.height +
+        positionRect.height - totalDurationOffset;
+
+    const topOffset = Math.abs(
+        positionRect.top - this.resolvedTopMargin_ -
+        (adjustedViewportRect.height -
+            (positionRect.height * this.bottomRatio_)
+        )
+      );
+
+    this.scrollProgress_ = topOffset / totalDuration;
+  }
+
+  /**
+   * @private
+   */
+  parseAttributes_() {
+    // Ratio is either "<top-bottom:{0,1}>" or "<top:{0,1}> <bottom:{0,1}>"
+    // e.g, "0.5 1": use 50% visibility at top and 100% at the bottom of viewport.
+    const ratios = this.element.getAttribute('intersection-ratios');
+    if (ratios) {
+      const topBottom = ratios.trim().split(' ');
+      this.topRatio_ = this.validateAndResolveRatio_(topBottom[0]);
+      this.bottomRatio_ = this.topRatio_;
+      if (topBottom[1]) {
+        this.bottomRatio_ = this.validateAndResolveRatio_(topBottom[1]);
+      }
+    }
+
+    // Margin is either "<top-bottom:{px,vh}>" or "<top:{px,vh}> <bottom:{px,vh}>"
+    // e.g, "100px 10vh": exclude 100px from top and 10vh from bottom of viewport.
+    const margins = this.element.getAttribute('exclusion-margins');
+    if (margins) {
+      const topBottom = margins.trim().split(' ');
+      this.topMarginExpr_ = topBottom[0];
+      this.bottomMarginExpr_ = this.topMarginExpr_;
+      if (topBottom[1]) {
+        this.bottomMarginExpr_ = topBottom[1];
+      }
+    }
+
+    this.targetSelector_ = this.element.getAttribute('target-selector');
+  }
+
+  /**
+   * Finds the container scene. Either parent of the component or specified by
+   * `target-selector` attribute.
+   * @private
+   */
+  discoverScene_() {
+    if (this.targetSelector_) {
+      this.scene_ = user().assertElement(
+          this.getAmpDoc().getRootNode().querySelector(this.targetSelector_),
+          'No element found with query selector:' + this.targetSelector_);
+    } else {
+      this.scene_ = this.element.parentNode;
+    }
+    // Hoist body to documentElement.
+    if (this.scene_.tagName == 'BODY') {
```

Comment:
It could be `<amp-body>` for SD polyfill. Maybe better do `this.scene_ == ampdoc.getBody()` or `dom.js:isBody(x)`? We are not super well protected from this now, but in fact there's only one violation of this rule right now which I'm fixing.